### PR TITLE
Dataset search with results having small memory footprints

### LINKED ISSUE #697: Problem: Why grid_spatial is not a search field
### Problem: Why `grid_spatial` is not a search field
The `search_fields` used by `index.datasets.search` functions include `native fields` (those that are native to table schema), so called `fixed fields (not dynamic in metadata type schema)`, and other search fields defined in `metadata_type`. None of these include `grid_spatial`, which is puzzling. Except `telemetry`, all other `metadata types` we know of has `grid_spatial`. The `fixed fields` don't seem to correlate with `required fields` (id, creation_dt, label, sources) in metadata type schema either (but seem to be present in all metadata types we have listed). The problem with this is that, while we can still access `grid_spatial` selectively through database layer, we loose query processing capabilities of various search functions (particularly `search_returning` )

### PR DESCRIPTION
**Pull request by ashoka1234**
### Reason for this pull request
Large scale `dataset aggregate functions` such as `extent` computes are greatly facilitated by memory
efficient result datasets. The current `dataset search` functions has number of limitations in addition to large size of dataset objects (by those search functions that return `Dataset` objects). One of the limitations is `extracting custom fields`. Such custom fields could be `derived fields` such as `extent`, `crs`
etc. which are typically implemented as property functions in the `Dataset` class. This PR implement a 
search query function that combines the power of `Dataset` object returning search functions with 
memory efficiency of select field extracting search functions by means of dynamically generated `DatasetLight` class. 

This PR also includes a search function that returns `time bounds` of a given product, which are useful
to organise time-sliced concurrent workflows.

### Memory savings
If we use pickling to approximate memory size of python objects, example memory savings if we only
wanted `id` and `extent` information per dataset, are as follows:
Using `python dill` library: `index.datasets.search(limit=2, **query)` would give us 43374 for the first dataset while `index.datasets.search_returning_datasets_light(field_names=('id', 'extent'), limit=2, **query)` would give us 603 thus giving us very substantial memory footprint reduction.
...

### Proposed changes
- Implemented new search query function having the interface 
`search_returning_datasets_light(self, field_names: tuple, custom_offsets=None, limit=None, **query)`: In addition to functions implemented in `datacube.index.datasets`, further search functions
added to `postgres API` that guarantee `per dataset results` by aggregating information in the `dataset_location` table.  
- Implemented time bounds search function with the interface `get_product_time_bounds(self, product: str)`

 - [x] Closes #697 #711 #715 
 - [x] Tests added / passed
 - [ ] Fully documented, including `docs/about/whats_new.rst` for all changes

### CODE REVIEW

File: datacube/drivers/postgres/_api.py [hunk 1]
```diff
@@ -535,6 +536,77 @@ def search_datasets(self, expressions,
                                                   select_fields, with_source_ids, limit)
         return self._connection.execute(select_query)
 
+    @staticmethod
+    def search_unique_datasets_query(expressions, select_fields, limit):
```
Comment (Kirill888) [inline datacube/drivers/postgres/_api.py:540]: Needs better doc string, what does this function do? What does "unique" imply here?

File: datacube/index/_datasets.py [hunk 2]
```diff
@@ -703,3 +703,216 @@ def search_eager(self, **query):
...
+        :param limit: Number of datasets returned per product.
+        :param query: key, value mappings of query that will be processed against metadata_types,
+                      product definitions and/or dataset table.
+        :return: A Dynamically generated DatasetLight (a subclass of namedtuple and possibly with
+        property functions).
+        """
+
+        assert field_names
+
+        class DatasetSpatialMixin(object):
```
Comment (Kirill888) [inline datacube/index/_datasets.py]: Does this class need to be here, inside the function? Rather take it out into module scope.

File: datacube/drivers/postgres/_api.py [hunk 3]
```diff
@@ -535,6 +536,77 @@ def search_datasets(self, expressions,
...
+            ).where(
+                where_expr
+            ).limit(
+                limit
+            )
+        )
+
+    def search_unique_datasets(self, expressions, select_fields=None, limit=None):
+        """
+        Processes a search query without duplicating datasets.
```
Comment (Kirill888) [inline datacube/drivers/postgres/_api.py]: needs better doc string, what does "unique" mean,  how is it different from just `search_datasets"

File: datacube/index/_datasets.py [hunk 4]
```diff
@@ -703,3 +703,222 @@ def search_eager(self, **query):
...
+        Parse and generate the list of select fields to be passed to the database API and
+        those fields that are to be further processed once the results are returned.
+        """
+
+        assert product and field_names
+
+        dataset_fields = product.metadata_type.dataset_fields
+        dataset_section = product.metadata_type.definition['dataset']
+
+        from datacube.drivers.postgres._fields import SimpleDocField
```
Comment (Kirill888) [inline datacube/index/_datasets.py]: please move imports to the top of the function, or better to module level

File: datacube/index/_datasets.py [hunk 5]
```diff
@@ -703,3 +703,222 @@ def search_eager(self, **query):
...
+                        return Dataset.bounds.__get__(self)
+
+            with self._db.connect() as connection:
+                results = connection.search_unique_datasets(
+                    query_exprs,
+                    select_fields=select_fields,
+                    limit=limit
+                )
+
+            import json
```
Comment (Kirill888) [inline datacube/index/_datasets.py]: import in odd location

File: datacube/drivers/postgres/_api.py [hunk 6]
```diff
@@ -535,6 +536,77 @@ def search_datasets(self, expressions,
...
+
+        # expressions involving 'uri' and 'uris' will be handled different
+        expressions = [expression for expression in expressions
+                       if expression.field.required_alchemy_table != DATASET_LOCATION]
+
+        if select_fields:
+            select_columns = []
+            for field in select_fields:
+                if field.name in {'uri', 'uris'}:
+                    # All active URIs, from newest to oldest
```
Comment (Kirill888) [inline datacube/drivers/postgres/_api.py:564]: this part is not being tested, please add tests for uri fields

File: datacube/index/_datasets.py [hunk 7]
```diff
@@ -703,3 +703,216 @@ def search_eager(self, **query):
         :rtype: list[Dataset]
         """
         return list(self.search(**query))
+
+    def get_product_time_bounds(self, product: str):
```
Comment (Kirill888) [inline datacube/index/_datasets.py:736]: this function is not covered by tests at all, please add some

File: datacube/index/_datasets.py [hunk 8]
```diff
@@ -703,3 +703,216 @@ def search_eager(self, **query):
...
+                    __slots__ = ()
+
+            with self._db.connect() as connection:
+                results = connection.search_unique_datasets(
+                    query_exprs,
+                    select_fields=select_fields,
+                    limit=limit
+                )
+
+            import json
```
Comment (Kirill888) [inline datacube/index/_datasets.py]: imports in odd location, move to top of function or module

**Review by Kirill888 — CHANGES_REQUESTED** [review]
need tests for `get_product_time_bounds` and for `uri(s)` field extraction + some cosmetics, see comments.

File: datacube/drivers/postgres/_api.py (hunk 1 above)
Comment (Kirill888) [reply datacube/drivers/postgres/_api.py:540]: after some discussions this is fine as is, it is meant to be internal method and "unique" refers to "not returning duplicates"

Comment (ashoka1234) [reply datacube/drivers/postgres/_api.py:540]: Done

File: datacube/index/_datasets.py (hunk 2 above)
Comment (ashoka1234) [reply datacube/index/_datasets.py]: Done

File: datacube/drivers/postgres/_api.py (hunk 3 above)
Comment (ashoka1234) [reply datacube/drivers/postgres/_api.py]: Done

File: datacube/index/_datasets.py (hunk 4 above)
Comment (ashoka1234) [reply datacube/index/_datasets.py]: Done

File: datacube/index/_datasets.py (hunk 5 above)
Comment (ashoka1234) [reply datacube/index/_datasets.py]: Moved to top of the module

File: datacube/drivers/postgres/_api.py (hunk 6 above)
Comment (ashoka1234) [reply datacube/drivers/postgres/_api.py:564]: Done

File: datacube/index/_datasets.py (hunk 7 above)
Comment (ashoka1234) [reply datacube/index/_datasets.py:736]: Done

File: datacube/index/_datasets.py (hunk 8 above)
Comment (ashoka1234) [reply datacube/index/_datasets.py]: Done

**Review by Kirill888 — APPROVED** [review]
thanks

**Comment by jeremyh** [comment]
I was working on a review but was too slow. I agree with Kirill's line-by-line comments, but I'm not a big fan of the change at a higher level, particularly the public api it adds.

Could we discuss this in person before we include this in the release? There was related discussion in #715 previously.

**Comment by Kirill888** [comment]
@jeremyh haven't noticed your comment until it was too late. Let's discuss this tomorrow, Friday, I should be in the office.

I agree that this is only a stop-gap measure. The api part is not very relevant in my opinion as we will be changing core<->db interface in the next release anyway, so this just some internal tooling that might aid in migration.

**Pull request merged**